vsphere/datadog_checks/vsphere/vsphere.py

```diff
@@ -231,34 +226,55 @@ def _get_server_instance(self, instance):
-                )
-            except Exception as e:
-                err_msg = "Connection to %s failed: %s" % (instance.get('host'), e)
-                self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.CRITICAL,
-                                   tags=service_check_tags, message=err_msg)
-                raise Exception(err_msg)
+        try:
+            # Object returned by SmartConnect is a ServerInstance
+            #   https://www.vmware.com/support/developer/vc-sdk/visdk2xpubs/ReferenceGuide/vim.ServiceInstance.html
+            server_instance = connect.SmartConnect(
+                host=instance.get('host'),
+                user=instance.get('username'),
+                pwd=instance.get('password'),
+                sslContext=context if not ssl_verify or ssl_capath else None
+            )
+        except Exception as e:
+            err_msg = "Connection to {} failed: {}".format(instance.get('host'), e)
+            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.CRITICAL,
+                               tags=service_check_tags, message=err_msg)
+            raise ConnectionError(err_msg)
+
+        # Check that we have sufficient permission for the calls we need to make
+        try:
+            server_instance.CurrentTime()
+        except Exception:
+            err_msg = (
+                "A connection to {} can be made, but it appears the user {} doesn't have enough permissions"
+            ).format(instance.get('host'), instance.get('username'))
+            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.CRITICAL,
+                               tags=service_check_tags, message=err_msg)
+            raise ConnectionError(err_msg)
+
+        return server_instance
+
+    def _get_server_instance(self, instance):
+        i_key = self._instance_key(instance)
+        tags = instance.get('tags', [])
+
+        service_check_tags = [
+            'vcenter_server:{0}'.format(instance.get('name')),
+            'vcenter_host:{0}'.format(instance.get('host')),
+        ] + tags
+        service_check_tags = list(set(service_check_tags))
 
-            self.server_instances[i_key] = server_instance
+        if i_key not in self.server_instances:
+            self.server_instances[i_key] = self._smart_connect(instance, service_check_tags)
 
         # Test if the connection is working
         try:
-            self.server_instances[i_key].RetrieveContent()
+            self.server_instances[i_key].CurrentTime()
             self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.OK,
                                tags=service_check_tags)
-        except Exception as e:
-            err_msg = "Connection to %s died unexpectedly: %s" % (instance.get('host'), e)
-            self.service_check(self.SERVICE_CHECK_NAME, AgentCheck.CRITICAL,
-                               tags=service_check_tags, message=err_msg)
-            raise Exception(err_msg)
+        except Exception:
```

Comment:
shouldn't we catch `ConnectionError` at this point?

Reply:
The exception here is the exception raised by `CurrentTime`, which is not a `ConnectionError`.